### PR TITLE
4767

### PR DESCRIPTION
**Pull request by emitche**
- Manage the association of question groups with loan types - front end

### CODE REVIEW

File: app/assets/javascripts/backbone/views/loan_questions_view.coffee [hunk 1]
```diff
@@ -34,13 +35,15 @@ class MS.Views.LoanQuestionsView extends Backbone.View
       @$('#edit-modal').modal('show')
       new MS.Views.TranslationsView(el: $('[data-content-translatable="custom_field"]'))
       @$('#custom_field_parent_id').val(parent_id)
+      $('.loan-types').select2({theme: "bootstrap"})
```
Comment (hooverlunch) [inline app/assets/javascripts/backbone/views/loan_questions_view.coffee]: These can/should be `@$` right?

File: app/assets/javascripts/backbone/views/loan_questions_view.coffee [hunk 2]
```diff
@@ -34,13 +35,15 @@ class MS.Views.LoanQuestionsView extends Backbone.View
...
     )
 
   editNode: (e) ->
     id = @$(e.target).closest('li').data('id')
     @$('#edit-modal .modal-content').load("/admin/loan_questions/#{id}/edit", =>
       @$('#edit-modal').modal('show')
       new MS.Views.TranslationsView(el: $('[data-content-translatable="custom_field"]'))
+      $('.loan-types').select2({theme: "bootstrap"})
```
Comment (hooverlunch) [inline app/assets/javascripts/backbone/views/loan_questions_view.coffee]: I checked if select2 supports default themes, and it does:

https://select2.github.io/options.html#setting-default-options

This will remove some duplication and ensure consistency.

Also remember that you don't need `{}` inside `()` in CoffeeScript, just like Ruby.

File: app/assets/stylesheets/admin/forms/_loan_questions_form.scss [hunk 3]
```diff
@@ -0,0 +1,7 @@
+.loan_questions {
+  .modal {
+    .hidden {
+      display: none;
+    }
+  }
+}
```
Comment (hooverlunch) [inline app/assets/stylesheets/admin/forms/_loan_questions_form.scss]: I think this would be ok to put in general.scss as it's a useful general style.

File: app/models/custom_field.rb [hunk 4]
```diff
@@ -2,17 +2,18 @@
...
+#  data_type             :string
+#  has_embeddable_media  :boolean          default(FALSE), not null
+#  id                    :integer          not null, primary key
+#  internal_name         :string
+#  migration_position    :integer
+#  override_associations :boolean          default(FALSE), not null
+#  parent_id             :integer
+#  position              :integer
+#  required              :boolean          default(FALSE), not null
+#  updated_at            :datetime         not null
```
Comment (hooverlunch) [inline app/models/custom_field.rb]: Blargh I hate when code tries to be cutsie and display things in tabular format like this and totally break diff review! All 12 of the above lines changed due to one actual change! 
</rant>

File: app/models/custom_field.rb [hunk 5]
```diff
@@ -77,10 +78,11 @@ def self.loan_questions(field_set = nil)
   #   and optional for all others
   # - If override is true and no records are present, all are optional
   # - If override is false, inherit from parent
+  # - Root nodes effectively have override always true
   # Note, loan type association records are ignored for questions without the 'override_assocations'
   # flag assigned.
   def required_for?(loan)
-    if override_associations
+    if override_associations || !parent_id
```
Comment (hooverlunch) [inline app/models/custom_field.rb]: Curious why `parent_id` is being checked directly here. Prefer using closure_tree methods like `child?`.

File: app/models/embeddable_media.rb [hunk 6]
```diff
@@ -2,14 +2,25 @@
...
+#  owner_type      :string
+#  sheet_number    :string
+#  start_cell      :string
+#  updated_at      :datetime         not null
+#  url             :string
+#  width           :integer
+#
+# Indexes
+#
+#  index_embeddable_media_on_owner_type_and_owner_id  (owner_type,owner_id)
```
Comment (hooverlunch) [inline app/models/embeddable_media.rb]: Not sure why this change is part of this PR. Remove?

File: app/views/admin/loan_questions/_form.slim [hunk 7]
```diff
@@ -16,9 +16,21 @@
 
     = f.input :data_type, as: :select, label: t('loan_questions.data_type'),
       collection: @data_types
-      
+
```
Comment (hooverlunch) [inline app/views/admin/loan_questions/_form.slim]: Does everyone have auto-truncate turned on in their editor? I'm still seeing whitespace like this showing up...

File: app/views/admin/loan_questions/_form.slim [hunk 8]
```diff
@@ -16,9 +16,21 @@
...
     = f.input :has_embeddable_media, label: t('loan_questions.has_embeddable_media')
 
+    label Required for which loan types?
```
Comment (hooverlunch) [inline app/views/admin/loan_questions/_form.slim]: I18n please

File: app/views/admin/loan_questions/_form.slim [hunk 9]
```diff
@@ -16,9 +16,21 @@
...
+
+    - if @loan_question.parent
+      = f.collection_radio_buttons :override_associations,
+        [[false, 'Inherits from parent'], [true, 'Overrides parent']], :first, :last,
```
Comment (hooverlunch) [inline app/views/admin/loan_questions/_form.slim]: I18n

File: app/views/admin/loan_questions/_form.slim [hunk 10]
```diff
@@ -16,9 +16,21 @@
...
+        item_wrapper_class: 'radio'
+
+    - hideLoanTypes = @loan_question.parent && !@loan_question.override_associations ? 'hidden': ''
+    .loan-types-container class="#{hideLoanTypes}"
+      = f.input_field :loan_type_ids, collection: @loan_type_options, multiple: true,
+        class: "loan-types"
```
Comment (hooverlunch) [inline app/views/admin/loan_questions/_form.slim]: Looks like this is not getting I18n'd either?

**Important**

SimpleForm gives you a super easy way to I18n a select like this. There are already some examples in `en.yml`:

```
simple_form:
    labels:
      project_step_move:
        move_type: "Which Date?"
        shift_subsequent: "Shift Others?"
    options:
      person:
        owning_division_role:
          admin: "Admin"
          member: "Member"
          none: "None"
      project_step_move:
        move_type:
          change_date: "Change step's scheduled date"
          mark_completed: "Mark step completed on this date"
```

See also https://github.com/plataformatec/simple_form#i18n

We should always be using this.

File: db/schema.rb [hunk 11]
```diff
@@ -110,14 +110,23 @@
...
+    t.string   "owner_type"
+    t.string   "sheet_number"
+    t.string   "start_cell"
     t.datetime "updated_at", null: false
     t.string   "url"
     t.integer  "width"
   end
 
+  add_index "embeddable_media", ["owner_type", "owner_id"], name: "index_embeddable_media_on_owner_type_and_owner_id", using: :btree
+
```
Comment (hooverlunch) [inline db/schema.rb]: Please remove this stuff from the PR also.

File: app/assets/stylesheets/admin/forms/_loan_questions_form.scss (hunk 3 above)
Comment (linesarefuzzy) [inline app/assets/stylesheets/admin/forms/_loan_questions_form.scss]: Actually, bootstrap provides this so it shouldn't be necessary.

File: app/models/custom_field.rb (hunk 5 above)
Comment (linesarefuzzy) [inline app/models/custom_field.rb]: Oh, that was me, I'll fix.

File: app/models/embeddable_media.rb (hunk 6 above)
Comment (emitche) [inline app/models/embeddable_media.rb]: Can do.

File: app/views/admin/loan_questions/_form.slim (hunk 7 above)
Comment (emitche) [inline app/views/admin/loan_questions/_form.slim]: If you see it, let me know. Atom has issues with Slim files, so though I have settings to see it when it happens, sometimes I miss it. Thanks and sorry about that.

File: db/schema.rb (hunk 11 above)
Comment (emitche) [inline db/schema.rb]: Ok.

File: app/views/admin/loan_questions/_form.slim (hunk 10 above)
Comment (emitche) [inline app/views/admin/loan_questions/_form.slim]: Loan type options are being translated in the database? Are we moving away from that?

Comment (emitche) [inline app/views/admin/loan_questions/_form.slim]: I spoke with Fuzzy about this. There are translations in the database, which can be accessed in rails console with `Loan.loan_type_options`. I am using that dataset with the translations in this spot.
